### PR TITLE
Add support for entity schemas

### PR DESCRIPTION
**Pull request by Romusk**
Closes: https://github.com/metarhia/metaschema/issues/161
Closes: https://github.com/metarhia/metaschema/issues/214

### CODE REVIEW

File: lib/validate.js [hunk 1]
```diff
@@ -177,7 +177,78 @@ const validateCategory = (
...
+  const { path = '', patch = false } = options;
+  const errors = [];
+  const schemaProps = new Set(Object.keys(schema));
+  const objectProps = new Set(Object.keys(instance));
+  const props = new Set([...schemaProps, ...objectProps]);
+  for (const prop of props) {
+    const isSchemaProp = schemaProps.has(prop);
+    const isObjectProp = objectProps.has(prop);
+    if (isObjectProp && !isSchemaProp) {
+      errors.push(new ValidationError('unresolvedProperty', `${path}${prop}`));
```
Comment (nechaido) [inline lib/validate.js]: You should also check for `Symbol.for('other')` if the schema has it then you should validate against it.

File: lib/config.js [hunk 2]
```diff
@@ -80,6 +80,55 @@ const postprocessCategory = (category, ms) => {
...
+        );
+      }
+    }
+  }
+  return errors;
+};
+
 const processOrder = {
   domains: 0,
   category: 1,
```
Comment (nechaido) [inline lib/config.js]: Add `entitySchema` here.

**Comment by nechaido** [comment]
I'd rename `validateCategory` to `validateSchema` and use it for both category and schemas.

File: lib/fs-loader.js [hunk 3]
```diff
@@ -143,6 +143,23 @@ const load = async (dir, options, config) => {
...
 };
 
+const asyncReadSchemas = async (path, systemParams, params) => {
+  const systemSchema = await load(
+    path,
+    systemParams.options,
+    systemParams.config
+  );
+  for (const type of Object.keys(params.config.processOrder)) {
+    if (systemSchema.schemas.filter(single => single.type === type)) {
```
Comment (nechaido) [inline lib/fs-loader.js]: Why not:
`if (systemSchemas.enitySchemas.has(type))`

File: lib/fs-loader.js [hunk 4]
```diff
@@ -143,6 +143,23 @@ const load = async (dir, options, config) => {
   return Metaschema.create(schemas, config);
 };
 
+const asyncReadSchemas = async (path, systemParams, params) => {
```
Comment (nechaido) [inline lib/fs-loader.js]:
```suggestion
const readConfig = async (path, systemParams, params) => {
```
What do you think @lundibundi @belochub?

File: lib/fs-loader.js [hunk 5]
```diff
@@ -143,6 +143,23 @@ const load = async (dir, options, config) => {
...
+  const systemSchema = await load(
+    path,
+    systemParams.options,
+    systemParams.config
+  );
+  for (const type of Object.keys(params.config.processOrder)) {
+    if (systemSchema.category.has(type)) {
+      params.config.processors[type].validate = [
```
Comment (nechaido) [inline lib/fs-loader.js]:
```suggestion
      params.config.processors[type].validateSchema = [
```

File: lib/validate.js [hunk 6]
```diff
@@ -122,6 +122,13 @@ const validateCategory = (
     const isSchemaProp = schemaProps.has(prop);
     const isObjectProp = objectProps.has(prop);
     if (isObjectProp && !isSchemaProp) {
+      if (
```
Comment (nechaido) [inline lib/validate.js]:
```suggestion
      if (schema[Symbol.for('other')])
```

Should work just fine.

File: lib/validate.js [hunk 7]
```diff
@@ -122,6 +122,70 @@ const validateCategory = (
...
+                );
+              }
+            }
+          } else {
+            errors.push(new ValidationError('unknownFormat', `${path}${prop}`));
+            continue;
+          }
+        }
+        continue;
+      }
```
Comment (nechaido) [inline lib/validate.js]:
```suggestion
      } else {
```

File: test/schemas/modules/Person/category.schema [hunk 8]
```diff
@@ -0,0 +1,3 @@
+{
+  [Symbol.for('other')]: ['FullName', {domain: 'CountryName'}, {category: 'FullName'}],
```
Comment (nechaido) [inline test/schemas/modules/Person/category.schema]:
```suggestion
  [Symbol.for('other')]: [
    'FullName',
    { domain: 'CountryName' },
    { category: 'FullName' },
  ],
```

You should also add a test to ensure that this schema works as expected.

File: test/schemas/utils/category.schema [hunk 9]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (nechaido) [inline test/schemas/utils/category.schema]: What's the purpose of this schema?

File: lib/validate.js [hunk 10]
```diff
@@ -122,6 +122,70 @@ const validateCategory = (
     const isSchemaProp = schemaProps.has(prop);
     const isObjectProp = objectProps.has(prop);
     if (isObjectProp && !isSchemaProp) {
+      if (schema[Symbol.for('other')]) {
+        let isValid;
+        for (const singleInstance of schema[Symbol.for('other')]) {
```
Comment (nechaido) [inline lib/validate.js]: This should be a lot shorter:
```js
const valid =
  schema[Symbol.for('other')].filter(type => {
    let kind, name;
    if (type.domain) {
      kind = 'domains';
      name = type.domain;
    } else {
      kind = 'category';
      name = typeof type === 'string' ? type : type.category;
    }
    return !!ms.validate(kind, name, instances[prop]);
  }).length === 0;
```

File: test/schemas/utils/category.schema (hunk 9 above)
Comment (Romusk) [reply test/schemas/utils/category.schema]: > What's the purpose of this schema?

Just to test validation function.

File: test/readConfig.js [hunk 11]
```diff
@@ -0,0 +1,50 @@
+'use strict';
+
+const metatests = require('metatests');
+
+const { clone } = require('@metarhia/common');
+
+const {
+  // eslint-disable-next-line
```
Comment (nechaido) [inline test/readConfig.js]: What is the specific rule you need disabled?

File: lib/validate.js [hunk 12]
```diff
@@ -124,7 +124,40 @@ const validateCategory = (
...
+          schema[Symbol.for('other')].filter(type => {
+            let kind, name;
+            if (type.domain) {
+              kind = 'domains';
+              name = type.domain;
+            } else {
+              kind = 'category';
+              name = typeof type === 'string' ? type : type.category;
+            }
+            return !!ms.validate(kind, name, instance[prop], {});
```
Comment (nechaido) [inline lib/validate.js]:
```suggestion
            return !!ms.validate(kind, name, instance[prop], { path: propPath, patch });
```

File: lib/validate.js [hunk 13]
```diff
@@ -124,7 +124,40 @@ const validateCategory = (
...
+          }).length === 0;
+        if (!valid) {
+          errors.push(
+            new ValidationError(
+              'invalidAddProp',
```
Comment (nechaido) [inline lib/validate.js]: You should define serializer for this type in `lib/error`.

File: lib/validate.js [hunk 14]
```diff
@@ -124,7 +124,40 @@ const validateCategory = (
     const isSchemaProp = schemaProps.has(prop);
     const isObjectProp = objectProps.has(prop);
     if (isObjectProp && !isSchemaProp) {
-      errors.push(new ValidationError('unresolvedProperty', `${path}${prop}`));
+      if (
+        Object.values(instance[prop]).filter(
```
Comment (nechaido) [inline lib/validate.js]: I don't see any value in doing so.

File: test/schemas/schemaValidation/systemSchema/categoryField.category [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+  parse: { domain: 'Function' },
+  check: { domain: 'Function' },
+
+  Check: Validate(instance => {
+    const isDomain = !!instance.domain;
+    const isCategory = !!instance.category;
+
+    return isDomain !== isCategory;
+  }),
+}
```
Comment (nechaido) [inline test/schemas/schemaValidation/systemSchema/categoryField.category]: No newline at the end of a file.

File: lib/errors.js [hunk 16]
```diff
@@ -71,6 +71,8 @@ ValidationError.serializers = {
     `Invalid value of a enum in a property '${error.property}' ` +
     `allowed: ${wrap(error.info.expected)}, actual: '${error.info.actual}'`,
   undefinedEntity: error => `Undefined ${error.info} '${error.property}'`,
+  invalidAddProp: error =>
+    `Property '${error.property}' can't be found in validation schema.`,
```
Comment (nechaido) [inline lib/errors.js]: Please change error message template so that reported message is similar to:
`'SchemaName.Property' must be a valid instance of one of domain 'A', category 'B', ...`.

File: test/default.js [hunk 17]
```diff
@@ -16,9 +16,9 @@ const domains = {
   DateTime: { type: 'object', class: 'Date' },
 };
 
-const FullName = {
+const fullName = {
```
Comment (nechaido) [inline test/default.js]: What is the reason behind this change?

File: test/readConfig.js [hunk 18]
```diff
@@ -0,0 +1,81 @@
+'use strict';
+
+const metatests = require('metatests');
+
+const { clone } = require('@metarhia/common');
+
+const {
+  /* eslint-disable no-unused-vars */
```
Comment (nechaido) [inline test/readConfig.js]:
```suggestion
  /* eslint-disable-next-line no-unused-vars */
```

File: test/readConfig.js [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+  default: { decorators, ...defParam },
+  fs: { load, readConfig },
+} = require('../metaschema');
+
+const { getSchemaDir } = require('./utils');
+
+metatests.test(
+  'Must properly validate application schemas by system schemas.',
```
Comment (nechaido) [inline test/readConfig.js]:
```suggestion
  'Must properly validate application schemas by system schemas',
```

File: test/readConfig.js [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+      true
+    );
+
+    test.strictSame(ms.domains.size, 8);
+    test.strictSame(ms.categories.size, 5);
+
+    test.strictSame(modulesMs.domains.size, 3);
+    test.strictSame(modulesMs.categories.size, 2);
+
+    test.strictSame(
```
Comment (nechaido) [inline test/readConfig.js]: I do not see a reason to test instance validation here, it is already tested in a dedicated test.

File: lib/validate.js [hunk 21]
```diff
@@ -124,7 +124,37 @@ const validateCategory = (
...
+        const valid = schema[Symbol.for('other')].some(type => {
+          let kind, name;
+          if (type.domain) {
+            kind = 'domains';
+            name = type.domain;
+          } else {
+            kind = 'category';
+            name = typeof type === 'string' ? type : type.category;
+          }
+          if (
```
Comment (nechaido) [inline lib/validate.js]:
```suggestion
          return !ms.validate(kind, name, instance[prop], { path: propPath, patch });
```

instead of lines 138-142.

File: test/readConfig.js [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+  default: { decorators, ...defParam },
+  fs: { load, readConfig },
+} = require('../metaschema');
+
+const { getSchemaDir } = require('./utils');
+
+metatests.test(
+  'Must properly validate application schemas by system schemas',
+  async test => {
+    /* eslint-disable-next-line no-unused-vars */
```
Comment (nechaido) [inline test/readConfig.js]: Why do we need this rule disabled here?

File: test/readConfig.js [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+      updatedParam = await readConfig(path, systemParam, param);
+      ms = await load(aPath, updatedParam.options, updatedParam.config);
+      modulesMs = await load(mPath, updatedParam.options, updatedParam.config);
+    } catch (error) {
+      test.fail(error);
+      test.end();
+      return;
+    }
+
+    try {
```
Comment (nechaido) [inline test/readConfig.js]: I'd prefer to extract this to a different subtest.

File: test/readConfig.js [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+      incorrectMs = await load(
+        iPath,
+        updatedParam.options,
+        updatedParam.config
+      );
+      test.fail();
+    } catch (err) {
+      test.throws(err);
```
Comment (nechaido) [inline test/readConfig.js]:
```suggestion
      test.strictSame(err, expetedError);
```

File: test/readConfig.js [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    test.strictSame(
+      updatedParam.config.processors.domains.hasOwnProperty('validateInstance'),
```
Comment (nechaido) [inline test/readConfig.js]: AFAIU, you should check that `validateSchema` exists not `validateInstance`.

File: test/readConfig.js [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
+      true
+    );
+    test.strictSame(
+      updatedParam.config.processors.category.hasOwnProperty(
+        'validateInstance'
```
Comment (nechaido) [inline test/readConfig.js]: Ditto.

File: test/readConfig.js [hunk 27]
```diff
@@ -0,0 +1,68 @@
...
+      ),
+      true
+    );
+
+    test.strictSame(ms.domains.size, 2);
```
Comment (nechaido) [inline test/readConfig.js]: This test is redundant, `test/load` already checks that schemas are loaded correctly.

File: test/schemas/incorrectApplicationSchema/Application/Computer.category [hunk 28]
```diff
@@ -0,0 +1,7 @@
+{
+    Processor: { domain: 'String' },
+    RAM: { domain: 'Number' },
+    HDD: { domain: 'Number' },
+    Model: 'Acer',
```
Comment (nechaido) [inline test/schemas/incorrectApplicationSchema/Application/Computer.category]: I'd prefer not to use any registered trademark.

File: test/schemas/schemaValidationSimple/systemSchema/categoryFieldCategory.category [hunk 29]
```diff
@@ -0,0 +1,3 @@
+{
+    category: { domain: 'String' },
+}
```
Comment (nechaido) [inline test/schemas/schemaValidationSimple/systemSchema/categoryFieldCategory.category]: Add newline.

File: test/schemas/schemaValidation/systemSchema/domain.category [hunk 30]
```diff
@@ -0,0 +1,20 @@
...
+  comment: { domain: 'String' },
+  control: { domain: 'String' },
+  unit: { domain: 'String' },
+  base: { domain: 'String' },
+  symbol: { domain: 'String' },
+  format: { domain: 'String' },
+  values: { domain: 'Array' },
+  enum: { domain: 'String' },
+  mul: { domain: 'Number' },
+  sensitive: { domain: 'Boolean' },
```
Comment (nechaido) [inline test/schemas/schemaValidation/systemSchema/domain.category]: Please remove irrelevant schema entires.

File: lib/errors.js [hunk 31]
```diff
@@ -71,6 +71,15 @@ ValidationError.serializers = {
     `Invalid value of a enum in a property '${error.property}' ` +
     `allowed: ${wrap(error.info.expected)}, actual: '${error.info.actual}'`,
   undefinedEntity: error => `Undefined ${error.info} '${error.property}'`,
+  invalidAddProp: error =>
+    `'${error.property}' must be a valid instance of one of` +
+    error.info.symbolArray.map(symbolElement =>
```
Comment (nechaido) [inline lib/errors.js]: I'd rename `symbolArray` into `allowed`.

File: package-lock.json [hunk 32]
```diff
@@ -5,17 +5,17 @@
   "requires": true,
   "dependencies": {
     "@babel/cli": {
-      "version": "7.2.0",
```
Comment (nechaido) [inline package-lock.json]: Please remove this file from PR.

File: test/readConfig.js [hunk 33]
```diff
@@ -0,0 +1,70 @@
...
+    test.strictSame(
+      updatedParam.config.processors.category.hasOwnProperty('validateSchema'),
+      true
+    );
+
+    test.end();
+  }
+);
+
+metatests.test('Must successfuly fail schema load', async test => {
```
Comment (nechaido) [inline test/readConfig.js]:
```suggestion
metatests.test('Must correctly validate schemas', async test => {
```

File: test/readConfig.js [hunk 34]
```diff
@@ -0,0 +1,70 @@
...
+  const systemParam = clone(defParam);
```
Comment (nechaido) [inline test/readConfig.js]: I don't see any reason in cloning params.

File: test/schemas/incorrectApplicationSchema/Application/Computer.category [hunk 35]
```diff
@@ -0,0 +1,7 @@
...
+    IncorrectField: { domain: 'IncorrectDomain' },
```
[the remark this answers is not in the document]
Comment (nechaido) [reply test/schemas/incorrectApplicationSchema/Application/Computer.category]: > You should also add an actual invalid property to test its validation.

This part of a comment is still not resolved.

File: lib/validate.js [hunk 36]
```diff
@@ -124,7 +124,34 @@ const validateCategory = (
     const isSchemaProp = schemaProps.has(prop);
     const isObjectProp = objectProps.has(prop);
     if (isObjectProp && !isSchemaProp) {
-      errors.push(new ValidationError('unresolvedProperty', `${path}${prop}`));
+      if (schema[Symbol.for('other')]) {
+        const propPath = `${path}${prop}`;
+        const valid = schema[Symbol.for('other')].some(type => {
+          let kind, name;
```
Comment (bugagashenkj) [inline lib/validate.js]: IMHO you need to rename the type and kind to something more meaningful. Means the same thing.

File: test/readConfig.js [hunk 37]
```diff
@@ -0,0 +1,70 @@
...
+
+    const path = getSchemaDir('schemaValidationSimple');
+    const aPath = getSchemaDir('applicationSchema');
+    const mPath = getSchemaDir('modules');
+
+    const updatedParam = await readConfig(path, systemParam, param);
+    await load(aPath, updatedParam.options, updatedParam.config);
+    await load(mPath, updatedParam.options, updatedParam.config);
+
+    test.strictSame(
```
Comment (bugagashenkj) [inline test/readConfig.js]:
```suggestion
    test.assert(updatedParam.config.processors.domains.hasOwnProperty('validateSchema'));
```

File: test/readConfig.js [hunk 38]
```diff
@@ -0,0 +1,70 @@
...
+      updatedParam.config.processors.domains.hasOwnProperty('validateSchema'),
+      true
+    );
+    test.strictSame(
```
Comment (bugagashenkj) [inline test/readConfig.js]: Ditto

File: lib/fs-loader.js [hunk 39]
```diff
@@ -143,6 +143,121 @@ const load = async (dir, options, config) => {
...
+//            schema <Schema>
+//            instance <any>
+//            options <Object>
+//            Returns: <Error[]>
+//      processOrder: <Function> | <Object>, function is passed to
+//          Array.prototype.sort (a: <Schema>, b: <Schema>) => <number>.
+//          If <Object> is provided it would be used as
+//          map from schema type (<string>) to order value (<number>),
+//          types with lower values are processed earlier.
+// Returns: <workParam>
```
Comment (bugagashenkj) [inline lib/fs-loader.js]: IMHO you must add type.

File: test/readConfig.js [hunk 40]
```diff
@@ -17,8 +15,8 @@ const { getSchemaDir } = require('./utils');
 metatests.test(
   'Must properly validate application schemas by system schemas',
   async test => {
-    const systemParam = clone(defParam);
-    const param = clone(defParam);
+    const systemParam = defParam;
```
Comment (nechaido) [inline test/readConfig.js]: Why do we need a new variable?

File: test/readConfig.js [hunk 41]
```diff
@@ -28,22 +26,20 @@ metatests.test(
...
     );
 
     test.end();
   }
 );
 
 metatests.test('Must successfuly fail schema load', async test => {
-  const systemParam = clone(defParam);
-  const param = clone(defParam);
+  const systemParam = defParam;
```
Comment (nechaido) [inline test/readConfig.js]: Ditto.

File: lib/fs-loader.js [hunk 42]
```diff
@@ -143,6 +143,121 @@ const load = async (dir, options, config) => {
...
+const readConfig = async (path, systemParams, params) => {
+  const workParam = common.clone(params);
+  const systemSchema = await load(
+    path,
+    systemParams.options,
+    systemParams.config
+  );
+  for (const type of Object.keys(workParam.config.processOrder)) {
+    if (systemSchema.categories.has(type)) {
+      workParam.config.processors[type].validateSchema = [
```
Comment (nechaido) [inline lib/fs-loader.js]: We need to check, whether such an array already exists.

File: lib/fs-loader.js [hunk 43]
```diff
@@ -143,6 +143,121 @@ const load = async (dir, options, config) => {
...
+//            instance <any>
+//            options <Object>
+//            Returns: <Error[]>
+//      processOrder: <Function> | <Object>, function is passed to
+//          Array.prototype.sort (a: <Schema>, b: <Schema>) => <number>.
+//          If <Object> is provided it would be used as
+//          map from schema type (<string>) to order value (<number>),
+//          types with lower values are processed earlier.
+// Returns: workParam <Object>
+const readConfig = async (path, systemParams, params) => {
```
Comment (lundibundi) [inline lib/fs-loader.js]: Perhaps
```suggestion
const applySystemConfig = async (path, systemParams, params) => {
```
?

File: lib/fs-loader.js [hunk 44]
```diff
@@ -143,6 +143,121 @@ const load = async (dir, options, config) => {
...
+    if (systemSchema.categories.has(type)) {
+      workParam.config.processors[type].validateSchema = [
+        schema => {
+          const error = systemSchema.validate(
+            'category',
+            type,
+            schema.definition,
+            {
+              path: `${schema.name}.`,
+            }
```
Comment (lundibundi) [inline lib/fs-loader.js]: I think we should extract this `options` object to a variable to be able to write this call in one line. This should make it a bit clearer.

File: package-lock.json [hunk 45]
```diff
@@ -869,31 +869,36 @@
       "version": "4.0.0",
       "resolved": "https://registry.npmjs.org/arr-diff/-/arr-diff-4.0.0.tgz",
       "integrity": "sha1-1kYQdP6/7HHn4VI1dhoyml3HxSA=",
-      "dev": true
+      "dev": true,
```
Comment (lundibundi) [inline package-lock.json]: I think these should be reverted.

File: lib/fs-loader.js [hunk 46]
```diff
@@ -143,6 +143,123 @@ const load = async (dir, options, config) => {
...
+  const workParam = common.clone(params);
+  const systemSchema = await load(
+    path,
+    systemParams.options,
+    systemParams.config
+  );
+  for (const type of Object.keys(workParam.config.processOrder)) {
+    if (systemSchema.categories.has(type)) {
+      if (!workParam.config.processors[type].validateSchema) {
+        workParam.config.processors[type].validateSchema = [
```
Comment (lundibundi) [inline lib/fs-loader.js]: It should append validator if one exists, i.e.:
```js
const processor =  workParam.config.processors[type];
const validators = processor.validateSchema || [];
validators.push(schema => { ... })
processor.validateSchema = validators;
```

File: lib/errors.js [hunk 47]
```diff
@@ -71,6 +71,13 @@ ValidationError.serializers = {
     `Invalid value of a enum in a property '${error.property}' ` +
     `allowed: ${wrap(error.info.expected)}, actual: '${error.info.actual}'`,
   undefinedEntity: error => `Undefined ${error.info} '${error.property}'`,
+  invalidAddProp: error =>
+    `'${error.property}' must be a valid instance of one of:` +
+    error.info.allowed.map(def =>
+      def.domain
+        ? ` domain '${def.domain}'`
+        : ` category '${def.category ? def.category : def}'`
```
Comment (lundibundi) [inline lib/errors.js]:
```suggestion
        : ` category '${def.category || def}'`
```

File: lib/errors.js [hunk 48]
```diff
@@ -71,6 +71,13 @@ ValidationError.serializers = {
     `Invalid value of a enum in a property '${error.property}' ` +
     `allowed: ${wrap(error.info.expected)}, actual: '${error.info.actual}'`,
   undefinedEntity: error => `Undefined ${error.info} '${error.property}'`,
+  invalidAddProp: error =>
```
Comment (lundibundi) [inline lib/errors.js]: Why `Add` though? Perhaps just `invalidProperty` (considering we already have `unresolvedProperty`, `missingProperty`)?

File: lib/validate.js [hunk 49]
```diff
@@ -142,7 +142,33 @@ const validateCategory = (
     const isSchemaProp = schemaProps.has(prop);
     const isObjectProp = objectProps.has(prop);
     if (isObjectProp && !isSchemaProp) {
-      errors.push(new ValidationError('unresolvedProperty', `${path}${prop}`));
+      if (schema[Symbol.for('other')]) {
```
Comment (lundibundi) [inline lib/validate.js]: Why `other`, perhaps `allowedProperties` or  `allowedPropertyTypes` would fit better?

File: lib/errors.js (hunk 48 above)
Comment (nechaido) [reply lib/errors.js]: `invalidAddProp` is for Invalid Additional Property, as it is only reported in cases of properties that are not defined in a schema, thus `invalidProperty` is not the best fit.
Though I don't like `invalidAddProp` either.

File: lib/validate.js (hunk 49 above)
Comment (nechaido) [reply lib/validate.js]: What about `@metarhia/additionallyAllowedProperties`?

File: lib/errors.js (hunk 48 above)
Comment (o-rumiantsev) [reply lib/errors.js]: Maybe better name is `forbiddenAdditionalProp` or `notAllowedAdditionalProperty`?
